### ISSUE TITLE
Defer Parinfer first run if buffer is read only

### ISSUE THREAD

**Issue by andreyorst** [justinbarclay/parinfer-rust-mode#20]
Since prompt deferring was added, I was trying to modify `parinfer-rust--check-for-indentation` to skip the check if buffer is read only and run it before real modification, e.g. if buffer is no longer read only run check for indentation before change. If buffer is read only running the check `'immediate`ly is meaningless, as it can't be changed anyway, and it should be deferred.

My attempt was:

```diff
index 0d8e166..3667b48 100644
--- a/parinfer-rust-mode.el
+++ b/parinfer-rust-mode.el
@@ -351,12 +351,16 @@ If a change is detected in the buffer, prompt the user to see if they still want
 `parinfer-rust-mode' enabled."
   (when (parinfer-rust--execute-change-buffer-p "paren")
     (setq-local parinfer-rust--disable nil)
-    (if (y-or-n-p
-         "Parinfer needs to modify indentation in this buffer to work.  Continue? ")
+    (if (and (not buffer-read-only)
+             (y-or-n-p
+              "Parinfer needs to modify indentation in this buffer to work.  Continue? "))
         (let ((parinfer-rust--mode "paren"))
           (parinfer-rust--execute))
-      (parinfer-rust-mode -1)))
-  (remove-hook 'before-change-functions #'parinfer-rust--check-for-indentation t))
+      (parinfer-rust-mode -1)
+      (message "Parinfer needs to modify indentation, but buffer '%s' is read only" (current-buffer))
+      (add-hook 'before-change-functions #'parinfer-rust--check-for-indentation t t)))
+  (unless buffer-read-only
+    (remove-hook 'before-change-functions #'parinfer-rust--check-for-indentation t)))
```

However for some reason, `buffer-read-only` variable is `nil` when Parinfer runs its indentation check.

I'm not sure when does `parinfer-rust--check-for-indentation` run, before `buffer-read-only` is set or this is just how Emacs works, and I can't get around this problem for several days. Any idea why this can happen?

**Comment by justinbarclay** [comment]
Hmm if immediate mode is turned on by default `parinfer-rust--check-for--indentation` runs as soon as parinfer-rust-mode is turned on. Whether that is before or after `buffer-read-only` state is set I can not say. From what little testing I've done buffer-read-only is always set to true before a mode is assigned to it.

Btw rather than overwriting that funcion, I recommend advising. It's closer to the Emacs way IMO.

``` elisp  
(advice-add 'parinfer-rust--check-for-indentation :around #'aorst/check-buffer-read-only)`

(defun aorst/check-buffer-read-only (orig-func &rest args)
  (when (not buffer-read-only)
   (apply orig-func args)))
```

Alternatively, you could try something like 
``` elisp
(use-package parinfer-rust-mode
 ...
  :config
 (setq-local parinfer-rust-check-before-enable (if buffer-read-only
                                                                    'defer
                                                                    'immediate)))
```

I haven't really tested either of these pieces of code though so you still might not have luck

**Comment by andreyorst** [comment]
> Btw rather than overwriting that funcion, I recommend advising. It's closer to the Emacs way IMO.

I was hoping to send PR with this change :)

>Alternatively, you could try something like
>
>```emacs-lisp
>(use-package parinfer-rust-mode
> ...
>  :config
> (setq-local parinfer-rust-check-before-enable (if buffer-read-only
>                                                   'defer
>                                                 'immediate)))
>```
>I haven't really tested the first solution, but in a quick test of the second one it work well.

I'm not sure if this will work, as the variable will be set only once, no? Or `setq-local` has semantic of running for every new buffer?

**Comment by justinbarclay** [comment]
You're right... for some reason last night I was thinking :config ran every time the mode was enabled. My testing last night just lucked out because I first ran it on a read only buffer.

**Comment by andreyorst** [comment]
No problem :)

What do you think about this semantic in general? Enabling Parinfer for read only buffers seem unnecessary, as it is unlikely for such buffer to become writeable, but deferring Parinfer until first change may be nice option still, as there are rare occasions where we make buffers writeable, like in Ediff.

<sup>You really should sleep well btw, undersleeping or sleep deprivation is a dangerous habit!</sup>

**Comment by justinbarclay** [comment]
I had some issues with this idea and it took me a bit to realize what it was. I have hesitations on removing user agency and limiting when they want `parinfer-rust-mode` to run.  Now that I am able to voice my hesitation I can agree with you that this idea overall improves the experience for users.

 There is no need to prompt users to change the buffer when they can't make a change to it. Things get worse because then the buffer will be in an incorrect state if it becomes writable.

**Comment by justinbarclay** [comment]
I've got a branch here, in my small test case it worked well:

https://github.com/justinbarclay/parinfer-rust-mode/tree/handle-read-only-buffers

**Comment by andreyorst** [comment]
I've tested branch a bit, but i don't see the difference from `master` if using `'immediate` Parinfer prompts for read-only buffers as usual. Am I missing something?

**Comment by justinbarclay** [comment]
That is weird, on my installation of Arch linux I created a test file with only read permissions
``` clojure
   (def thing []
(hello))
```
And I only get a prompt when I try to edit the buffer after I disable read-only-mode.
https://asciinema.org/a/dTZvxbrRiNbcDO8DtkvSX8BGp

**Comment by andreyorst** [comment]
Hm, I'm jumping to Clojure source and immediately get the prompt:

https://asciinema.org/a/22ZJ5LwZbl2fqu5qwDvMveWkn

Here's what I did in recording:

- start CIDER repl
- set variable to `'immideate`
- jump to `defn` source code
  - parinfer kicks in, but buffer is read only
  - kill buffer
- set variable to `'defer`
- jump to `defn` source code
  - nothing happens (as it should)
  - toggle read only mode
  - try edit, parinfer kicks in (as it should)

**Comment by justinbarclay** [comment]
Ahh I see what you mean now. So what is happening here is that CIDER is opening up a file that's isn't actually read only but is enforcing a read only restriction on it as part of the cider-find-var function call. I am not sure how much more power parinfer-rust-mode has in this situation. It has no control over when it is enabled. Short of having users write protect clojure core files/folder I am not sure how to ensure parinfer-rust-mode always doing the correct things in this situation.

However, there are still ways to get this behavior on a _per project_ basis. If in your config file, you have `(setq parinfer-rust-check-before-enable 'defer)`

You could then add a `.dir-locals.el` file in each project and add this line:
``` elisp
((nil . ((parinfer-rust-check-before-enable . immediate))))
```
Then every file opened in that project behaves as if `parinfer-rust-check-before-enable` is set to `immediate` but anything outside of the project folder behaves as if it's set to `defer`.

You could even add the `.dir-locals.el` higher up in the directory tree to cover more than one project at once, but you have to make sure that those folder don't also contain files that should be in defer mode (in this case clojure core files).

For example... I have all parinfer-rust-mode in `/home/Justin/dev/parinfer-rust-mode`. So if this was the only project I wanted to have in immediate mode I could have `/home/Justin/dev/parinfer-rust-mode/.dir-locals.el`. But I have all my projects under the `dev` directory. So then I could have `/home/Justin/dev/.dir-locals.el`.  However, I would not want to put this file in my home directory, `/home/Justin/.dir-locals.el`, because then it would cause `immediate`  option to be applied to files in the .m2 folder as well.

**Comment by andreyorst** [comment]
>So what is happening here is that CIDER is opening up a file that's isn't actually read only but is enforcing a read only restriction on it as part of the cider-find-var function call. 

Oh, I didn't expect that. I thought that since sources are zip compressed they can't be edited, and it is Emacs who enforces readonliness. Now I see why my approach didn't work in the very beginning 😂

**Comment by justinbarclay** [comment]
Is there anything more that can be done here? If not, do you mind if I close this? As I’m not sure what more I can do for this particular issue.

**Comment by andreyorst** [comment]
yeah, I think that I got used to this behavior too. closing as solved

Thank you, you rock!)
